### PR TITLE
Update index table when generate service level readme

### PR DESCRIPTION
**Pull request by sima-zhu**
Python `azure-developer-loadtesting` package did not show up in preview page. After checkup, I found our index table is lack of ability of update.
Use the PR to fix the issue: https://github.com/Azure/azure-sdk-tools/issues/4646. 

Also fixed the issue: https://github.com/Azure/azure-sdk-tools/issues/4002

Testing pipeline: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1971665&view=results

Service readme update: https://github.com/MicrosoftDocs/azure-docs-sdk-python/commit/fa4a19f7cc31b0faf2fb66aa060de3a2eea0d6a9

Python `load-testing` preview service readme update: https://github.com/MicrosoftDocs/azure-docs-sdk-python/commit/fa4a19f7cc31b0faf2fb66aa060de3a2eea0d6a9#diff-e72fa81fcf60a88d964ba37bab8f4118f8dff8b3db96758b98e20031796681e1

### CODE REVIEW

File: eng/common/scripts/Helpers/Metadata-Helpers.ps1 [hunk 1]
```diff
@@ -94,14 +94,14 @@ function GetDocsMsService($packageInfo, $serviceName)
...
+  $updatedKeys = $updated | ForEach-Object { $_[0]}
   if (!$original) {
     return $updateMetdata 
   }
   $originalTable = ConvertFrom-StringData -StringData $original -Delimiter ":"
   foreach ($key in $originalTable.Keys) {
-    if (!($updated.ContainsKey($key))) {
+    if (!($updatedKeys.Contains($key))) {
       Write-Warning "New metadata missed the entry: $key. Adding back."
       $updateMetdata += "$key`: $($originalTable[$key])`r`n"
```
Comment (benbp) [inline eng/common/scripts/Helpers/Metadata-Helpers.ps1:106]:
```suggestion
  $updateMetadata = (($updated | ForEach-Object { "$($_[0]): $($_[1])" }) -join "`r`n") + "`r`n"
  $updatedKeys = $updated | ForEach-Object { $_[0]}
  if (!$original) {
    return $updateMetadata 
  }
  $originalTable = ConvertFrom-StringData -StringData $original -Delimiter ":"
  foreach ($key in $originalTable.Keys) {
    if (!($updated.ContainsKey($key))) {
    if (!($updatedKeys.Contains($key))) {
      Write-Warning "New metadata missed the entry: $key. Adding back."
      $updateMetadata += "$key`: $($originalTable[$key])`r`n"
```

File: eng/common/scripts/Helpers/Metadata-Helpers.ps1 [hunk 2]
```diff
@@ -129,18 +129,8 @@ function GenerateDocsMsMetadata($originalMetadata, $language, $languageDisplayNa
     $msauthor = $author
   }
   $date = Get-Date -Format "MM/dd/yyyy"
-
-  $metadataTable = @{
```
Comment (weshaggard) [inline eng/common/scripts/Helpers/Metadata-Helpers.ps1:133]: Have you considered [Ordered] here? https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_hash_tables?view=powershell-7.2#creating-ordered-dictionaries

**Comment by kurtzeborn** [comment]
While I see you're running this through the pipeline several times to validate that it works in an ad-hoc fashion, shouldn't you also have actual tests for this to be able to validate this again later when we come back to it for future changes?

File: eng/common/scripts/Helpers/Metadata-Helpers.ps1 (hunk 2 above)
Comment (sima-zhu) [reply eng/common/scripts/Helpers/Metadata-Helpers.ps1:133]: I don't want to order by alphabet. I want to have custom defined order.

Comment (weshaggard) [reply eng/common/scripts/Helpers/Metadata-Helpers.ps1:133]: That is what the [ordered] tag does.

File: eng/common/scripts/Service-Level-Readme-Automation.ps1 [hunk 3]
```diff
@@ -42,51 +42,64 @@ param(
 
 Set-StrictMode -Version 3
 
-function create-metadata-table($readmeFolder, $readmeName, $moniker, $msService, $clientTableLink, $mgmtTableLink, $serviceName)
+# Parse out the rest content of service readme after excluding index table and metadata.
+function Get-RestContent($readmePath) {
+  $restContent = ''
+  if (Test-Path $readmePath) {
+    $readmeContent = Get-Content $readmePath -Raw
```
Comment (weshaggard) [inline eng/common/scripts/Service-Level-Readme-Automation.ps1:49]: It is kind of difficult to understand what this is doing. Can you please write some tests so we can see that expected input and output to ensure it is doing what we expect.

File: eng/common/scripts/Service-Level-Readme-Automation.ps1 [hunk 4]
```diff
@@ -42,51 +42,64 @@ param(
...
-  Add-Content -Path $readmePath -Value $readmeHeader
-  Add-Content -Path $readmePath -Value $content -NoNewline
-}
-
-# Update the metadata table.
-function update-metadata-table($readmeFolder, $readmeName, $serviceName, $msService)
-{
-  $readmePath = Join-Path $readmeFolder -ChildPath $readmeName
-  $readmeContent = Get-Content -Path $readmePath -Raw
-  $match = $readmeContent -match "^---\n*(?<metadata>(.*\n?)*?)---\n*(?<content>(.*\n?)*)"
```
Comment (weshaggard) [inline eng/common/scripts/Service-Level-Readme-Automation.ps1:78]: Based on what I'm seeing you are trying to now exclude the table with the links from the content here. Why do we need to do that? I thought that was exactly why we split the package table into another MD that folks could include if they want. If they don't do the include then they won't get the table.

File: eng/common/scripts/Service-Level-Readme-Automation.ps1 [hunk 5]
```diff
@@ -42,19 +42,16 @@ param(
 
 Set-StrictMode -Version 3
 
-function create-metadata-table($readmeFolder, $readmeName, $moniker, $msService, $clientTableLink, $mgmtTableLink, $serviceName)
+function create-metadata-table($readmeFolder, $readmeName, $moniker, $msService, $indexTableLink, $mgmtTableLink, $serviceName)
```
Comment (weshaggard) [inline eng/common/scripts/Service-Level-Readme-Automation.ps1:45]: I assume you can remove the mgmtTableLink now as well.

File: eng/common/scripts/Service-Level-Readme-Automation-Functions.ps1 [hunk 6]
```diff
@@ -0,0 +1,90 @@
+
+function create-service-readme($readmeFolder, $readmeName, $moniker, $msService, $indexTableLink, $serviceName, $author, $msAuthor)
+{
+  $readmePath = Join-Path $readmeFolder -ChildPath $readmeName
```
Comment (weshaggard) [inline eng/common/scripts/Service-Level-Readme-Automation-Functions.ps1:4]: This functions are difficult to review given they were moved. Did you change anything in them?

**Review by weshaggard — APPROVED** [review]
One comment about the helpers script but otherwise looks good.

**Pull request merged**